[DerDocs Home](https://github.com/mcsimenc/DerLab/blob/master/DerDocsHome.md)

# Guide to MAKER Genome Annotation Pipeline on Kepler

### Overview
MAKER is a software pipeline written in Perl  designed for annotating whole-genome assemblies but it may also be useful for annotating shorter sequences. MAKER streamlines genome annotation by automatically carrying out processes such as sequence alignment, ab initio gene prediction, handling intermediate data files, and synthesizing final annotations from multiple lines of evidence. It is meant to be run iteratively, and can update and extend existing genome annotations, a process that works best if the existing genome annotation was created by MAKER. Protein-coding gene annotations are expected to be most complete if a combination of transcriptome, proteome, and HMM-based predictions are used. Repeat loci are annotated and some are classified by RepeatMasker if classifications are present in the repeat database used; Repbase is a good repeat database to include and this can be supplemented with a de novo set of consensus repeat sequences generated by RepeatModeler (separate from MAKER). tRNA, rRNA, and snoRNA can be identified as part of a MAKER run by turning on the relevant options and adding the relevant paths to maker_exe.ctl.

### Loading MAKER
MAKER is installed both locally (`/home/joshd/software/`) and globally (`/share/apps/genomics/`). The global installation is probably less reliable than the local one. If you don't already have one, you may need to create or copy a local module file for maker. Information about how to set up your module environment is in [KeplerModules.md](https://github.com/mcsimenc/DerLab/blob/master/KeplerModules.md)

To load MAKER, type

`module load local/maker`

A message should print with the MAKER version number and a short description. This information is hardcoded in the maker module file at 

### Setting up your MAKER space
It's best to create a new directory in which to run MAKER from. When ready to start the run, this directory will contain three control files, your PBS/torque/qsub/shell script to submit to the resource manager TORQUE using `qsub`, explained below. While running, MAKER will generate a very large number of files. Warning: moving a MAKER output directory around in your filesystem takes a very long time.

In this document, $RUNDIR is the location of the hypothetical directory from which MAKER will be executed.

#### Control Files
To generate template control files, from within $RUNDIR and after loading the MAKER module, type `maker -CTL`. This will generate three files:


[maker_opts.ctl](https://github.com/mcsimenc/DerLab/blob/master/maker_opts.md) contains general settings, \
[maker_bopts.ctl](https://github.com/mcsimenc/DerLab/blob/master/maker_bopts.md) contains BLAST settings, \
[maker_exe.ctl](https://github.com/mcsimenc/DerLab/blob/master/maker_exe.md) contains paths to dependency executables.

The syntax for the control files is key=value with # preceeded comments. The control files contain descriptive comments. Here is my explanation of the most important options to know about. Default BLAST settings were chosen because MAKER performed well under them for eukaryotic genomes.

DNA and protein sequences can be provided to MAKER in FASTA or GFF3 format. Below, the FASTA options are shown.

##### Notable options in maker_opts.ctl

###### syntax: `key=value`
###### e.g., `genome=mygenome.fasta`

##### Genome sequence in FASTA format
`genome`

##### eukaryotic (default) or prokaryotic genome.
`organism_type=eukaryotic`

##### Expressed sequences (e.g. EST, mRNA-seq transcripts) MAKER will use mappings of these sequences as as evidence for the existence of genes, and will directly infer models from genes using Exonerate est2genome if est2genome=1
`est`

##### Expressed sequences (e.g. EST, mRNA-seq transcripts) from a different organism than the one you are annotating. MAKER will use alignments of these sequences as as evidence for the existence of genes, and will directly infer models from genes using Exonerate est2genome if est2genome=1. This option will likley not add much to the annotation if assemble mRNA-seq data are given with `est`.
`altest`

##### Protein sequences. MAKER will use homology to these proteins as evidence for the existence of genes, and will directly infer models from genes using Exonerate protein2genome if protein2genome=1. The MAKER documentation suggests using "NP" sequences from ref-seq but not unreviewed sequences. 
`protein`

##### Repeat library. Repbase is a good start, but de novo libraries can be created with RepeatModeler.
rmlib= #provide an organism specific repeat library in fasta format for RepeatMasker

##### Transposable element proteins to assist with masking. A file containing some called `te_proteins.fasta` is included with MAKER at `makerinstalldir/data/`
`repeat_protein`

##### SNAP HMM file. A trained HMM for the gene predictor SNAP.
`snaphmm`

##### GeneMark HMM file. A trained HMM for the gene predictor GeneMark.
`gmhmm`

##### Augustus HMM file. A trained HMM for the gene predictor Augustus.
augustus_species= #Augustus gene prediction species model

##### FGENESH HMM file. A trained HMM for the gene predictor FGENESH.
`fgenesh_par_file`

#
est2genome=0 #infer gene predictions directly from ESTs, 1 = yes, 0 = no

protein2genome=0 #infer predictions from protein homology, 1 = yes, 0 = no

trna=0 #find tRNAs with tRNAscan, 1 = yes, 0 = no

snoscan_rrna= #rRNA file to have Snoscan find snoRNAs

This could be used to annotate protein-coding transposable elements.
unmask=0 #also run ab-initio prediction programs on unmasked sequence, 1 = yes, 0 = no

cpus=1 #max number of cpus to use in BLAST and RepeatMasker (not for MPI, leave 1 when using MPI)

pred_stats=0 #report AED and QI statistics for all predictions as well as models

min_protein=0 #require at least this many amino acids in predicted proteins

alt_splice=0 #Take extra steps to try and find alternative splicing, 1 = yes, 0 = no

always_complete=0 #extra steps to force start and stop codons, 1 = yes, 0 = no

split_hit=10000 #length for the splitting of hits (expected max intron size for evidence alignments)

single_exon=0 #consider single exon EST evidence when generating annotations, 1 = yes, 0 = no

single_length=250 #min length required for single exon ESTs if 'single_exon is enabled'

For re-annotating a genome, set to 1
map_forward=0 #map names and attributes forward from old GFF3 genes, 1 = yes, 0 = no

To save disk space
clean_up=0 #removes theVoid directory with individual analysis files, 1 = yes, 0 = no


##### Paths in maker_exe.ctl

```
#-----Location of Executables Used by MAKER/EVALUATOR
makeblastdb= #location of NCBI+ makeblastdb executable
blastn= #location of NCBI+ blastn executable
blastx= #location of NCBI+ blastx executable
tblastx= #location of NCBI+ tblastx executable
RepeatMasker=/home/joshd/software/maker/bin/../exe/RepeatMasker/RepeatMasker #location of RepeatMasker executable
exonerate=/home/joshd/software/maker/bin/../exe/exonerate/bin/exonerate #location of exonerate executable

#-----Ab-initio Gene Prediction Algorithms
snap=/home/joshd/software/maker/bin/../exe/snap/snap #location of snap executable
gmhmme3= #location of eukaryotic genemark executable
gmhmmp= #location of prokaryotic genemark executable
augustus= #location of augustus executable
fgenesh= #location of fgenesh executable
tRNAscan-SE= #location of trnascan executable
snoscan= #location of snoscan executable
```

#### Submission Script

```
#!/bin/bash
#PBS -k oe
#PBS -N JobName
#PBS -q q40
#PBS -j oe
#PBS -m ea
#PBS -M mcsimenc@csu.fullerton.edu
#PBS -l nodes=1:ppn=40

module load local/maker
module load local/trnascan
module load openmpi

cd /home/derstudent/data/santalales/annotation_all_taxa


echo "START"
date

mpirun -n 40 maker maker_bopts.ctl maker_exe.ctl maker_opts.ctl 1>maker.err 2>maker.log

echo "END"
date
```

[DerDocs Home](https://github.com/mcsimenc/DerLab/blob/master/DerDocsHome.md)
